core/src/main/scala/org/apache/spark/scheduler/ExecutorBlacklistTracker.scala

```diff
@@ -0,0 +1,175 @@
+ *    of this cluster.
+ *
+ * Also max number of blacklisted executors will not exceed the
+ * spark.scheduler.blacklist.maxBlacklistFraction of whole cluster, and blacklisted executors
+ * will be forgiven when there is no failure tasks in the
+ * spark.scheduler.blacklist.executorFaultTimeoutWindowInMinutes.
+ */
+private[spark] class ExecutorBlacklistTracker(conf: SparkConf) extends SparkListener {
+  import ExecutorBlacklistTracker._
+
+  private val maxBlacklistFraction = conf.getDouble(
+    "spark.scheduler.blacklist.maxBlacklistFraction", MAX_BLACKLIST_FRACTION)
+  private val avgBlacklistThreshold = conf.getDouble(
+    "spark.scheduler.blacklist.averageBlacklistThreshold", AVERAGE_BLACKLIST_THRESHOLD)
+  private val executorFaultThreshold = conf.getInt(
+    "spark.scheduler.blacklist.executorFaultThreshold", EXECUTOR_FAULT_THRESHOLD)
+  private val executorFaultTimeoutWindowInMinutes = conf.getInt(
+    "spark.scheduler.blacklist.executorFaultTimeoutWindowInMinutes", EXECUTOR_FAULT_TIMEOUT_WINDOW)
+
+  // Count the number of executors registered
+  var numExecutorsRegistered: Int = 0
+
+  // Track the number of failure tasks and time of latest failure to executor id
+  val executorIdToTaskFailures = new mutable.HashMap[String, ExecutorFailureStatus]()
+
+  // Clock used to update and exclude the executors which are out of time window.
+  private var clock: Clock = new SystemClock()
+
+  // Executor that handles the scheduling task
+  private val executor = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+    "spark-scheduler-blacklist-expire-timer")
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireTimeoutExecutorBlacklist())
+      }
+    }
+    executor.scheduleAtFixedRate(scheduleTask, 0L, 60, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    executor.shutdown()
+    executor.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  def setClock(newClock: Clock): Unit = {
+    clock = newClock
+  }
+
+  def getExecutorBlacklist: Set[String] = synchronized {
+    executorIdToTaskFailures.filter(_._2.isBlackListed).keys.toSet
+  }
+
+  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+    taskEnd.reason match {
+      case _: FetchFailed | _: ExceptionFailure | TaskResultLost |
+          _: ExecutorLostFailure | UnknownReason =>
+        val failureStatus = executorIdToTaskFailures.getOrElseUpdate(taskEnd.taskInfo.executorId,
+          new ExecutorFailureStatus)
+        failureStatus.numFailures += 1
+        failureStatus.updatedTime = clock.getTimeMillis()
+
+        // Update the executor blacklist
+        updateExecutorBlacklist()
+      case _ => Unit
+    }
+  }
+
+  override def onExecutorAdded(executorAdded: SparkListenerExecutorAdded): Unit = {
+    numExecutorsRegistered += 1
+  }
+
+  override def onExecutorRemoved(executorRemoved: SparkListenerExecutorRemoved
+      ): Unit = synchronized {
```

Comment:
I'm not totally sure if this style is OK or not ... but it does look weird to me.  I think probably best to just put the arg on its own line, as with normal multi-line function defs

```scala
override def onExecutorRemoved(
    executorRemoved: SparkListenerExecutorRemoved): Unit = synchronized {
```